File: scripts/build-containers.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Literal, Optional

ImageType = Literal['debug', 'production']


@dataclass
class ImageConfig:
    """Configuration for container images."""
    type: ImageType
    tag: str
    target: Optional[str] = None
    containerfile: Path = Path('Containerfile')


IMAGE_CONFIGURATION: dict[ImageType, ImageConfig] = {
    'debug': ImageConfig(
        type='debug',
        tag='local/bootstrap-olm:debug',
        target='debug',
    ),
    'production': ImageConfig(
        type='production',
        tag='local/bootstrap-olm:prod',
    ),
}


def read_go_version() -> str:
    """Read Go version from .go-version file."""
    with open('.go-version', 'r') as f:
        return f.read().strip()


def generate_build_command(
    config: ImageConfig,
    go_version: str,
) -> str:
    """Generate podman build command for given image type."""
    target_arg = f"--target {config.target} " if config.target else ''
    return (
        f"podman build --build-arg GO_VERSION={go_version} "
        f"{target_arg}"
        f"-t {config.tag} "
        f"-f {config.containerfile} ."
    )


def run_command(cmd: str) -> None:
    """Execute a shell command and handle errors."""
    try:
        subprocess.run(cmd, check=True, shell=True)
    except subprocess.CalledProcessError as e:
        print(
            f"Error running command: {cmd}",
            file=sys.stderr,
        )
        sys.exit(e.returncode)


def build_image(
    config: ImageConfig,
    go_version: str,
) -> None:
    """Build a single container image."""
    if not config.containerfile.exists():
        print(
            f"Error: Containerfile not found at {config.containerfile}",
            file=sys.stderr,
        )
        sys.exit(1)

    cmd = generate_build_command(config, go_version)
    print(f"Building {config.type} image...")
    run_command(cmd)
    print(f"Successfully built {config.type} image")


def main() -> None:
    """Main entry point."""
    go_version = read_go_version()
    print(f"Building images with Go version: {go_version}")

    for config in IMAGE_CONFIGURATION.values():
        build_image(config, go_version)


if __name__ == '__main__':
    main()

File: scripts/tasks/buildContainers.ts
#!/usr/bin/env yarn node

import { execSync } from "child_process";
import * as fs from "fs";
import * as toml from "toml";
import { getProjectRoot } from "../lib/project.ts";

type ImageType = "debug" | "production";
type BuildType = ImageType | "all";

interface ImageConfig {
  type: ImageType;
  tag: string;
  target?: string;
  containerfile: string;
}

function readImageConfiguration(): Record<ImageType, ImageConfig> {
  const configPath = `${getProjectRoot()}/containers.toml`;
  if (!fs.existsSync(configPath)) {
    console.error(`Error: Configuration file not found at ${configPath}`);
    process.exit(1);
  }

  const configContent = fs.readFileSync(configPath, "utf-8");
  const parsedConfig = toml.parse(configContent);

  // Validate and transform the TOML structure
  const config: Record<ImageType, ImageConfig> = {
    debug: {
      type: "debug",
      tag: parsedConfig.debug.tag,
      target: parsedConfig.debug.target,
      containerfile: parsedConfig.debug.containerfile,
    },
    production: {
      type: "production",
      tag: parsedConfig.production.tag,
      containerfile: parsedConfig.production.containerfile,
    },
  };

  return config;
}

function readGoVersion(): string {
  return execSync("mise current go", { encoding: "utf8" }).trim();
}

function generateBuildCommand(config: ImageConfig, goVersion: string): string {
  const targetArg = config.target ? `--target ${config.target} ` : "";
  return (
    `podman build --build-arg GO_VERSION=${goVersion} ` +
    `${targetArg}` +
    `-t ${config.tag} ` +
    `-f ${config.containerfile} .`
  );
}

function runCommand(cmd: string): void {
  try {
    execSync(cmd, { stdio: "inherit" });
  } catch (error) {
    console.error(`Error running command: ${cmd}`);
    if (error instanceof Error) {
      process.exit(1);
    }
    process.exit(error?.status || 1);
  }
}

function buildImage(config: ImageConfig, goVersion: string): void {
  if (!fs.existsSync(config.containerfile)) {
    console.error(`Error: Containerfile not found at ${config.containerfile}`);
    process.exit(1);
  }

  const cmd = generateBuildCommand(config, goVersion);
  console.log(`Building ${config.type} image...`);
  runCommand(cmd);
  console.log(`Successfully built ${config.type} image`);
}

function main(): void {
  const buildType = (process.argv[2] || "all") as BuildType;

  if (!["debug", "production", "all"].includes(buildType)) {
    console.error(
      "Invalid build type. Must be one of: debug, production, or all",
    );
    process.exit(1);
  }

  const goVersion = readGoVersion();
  console.log(`Building images with Go version: ${goVersion}`);

  const IMAGE_CONFIGURATION = readImageConfiguration();

  if (buildType === "all") {
    Object.values(IMAGE_CONFIGURATION).forEach((config) => {
      buildImage(config, goVersion);
    });
  } else {
    const config = IMAGE_CONFIGURATION[buildType];
    if (!config) {
      console.error(`Invalid build type: ${buildType}`);
      process.exit(1);
    }
    buildImage(config, goVersion);
  }
}

main();
